### PR TITLE
Fix/infinite loop on restub invoke

### LINKED ISSUE #31: Potential infinite loop on re-stubbing undefined global function without any return value
If we stub a previously undefined global function, [a new global function is generated](https://github.com/sirbrillig/spies/blob/master/src/Spies/GlobalSpies.php#L85).
And then if we re-stub it again without any return value, it gets redefined and an entry is created in `GlobalSpies::$redefined_functions` (because of changes in [this pr](https://github.com/sirbrillig/spies/pull/28)). Previously, the re-stub wouldn't work and there would be no entry in `GlobalSpies::$redefined_functions`.

But a new problem arises when the re-stub is invoked: `\Spies\GlobalSpies::handle_call_for` is called and in turn `\Spies\GlobalSpies::call_original_global_function` gets called (since stubbed without a return value).
Since there is an entry for the invoked function in `GlobalSpies::$redefined_functions`, the previous definition is restored (which happens to be the generated function) and `call_user_func_array` is called on that. This invokes `\Spies\GlobalSpies::handle_call_for` again and this keeps happening in a cycle.

Previously (before [this pr](https://github.com/sirbrillig/spies/pull/28)), the re-stub wouldn't work and thus the redefinition would never trigger. Only the first generated function or the original global function ([had it existed](https://github.com/sirbrillig/spies/issues/27)) would be invoked over and over again. In our case, the first generated function calls `\Spies\GlobalSpies::handle_call_for` like above, but it [returns early from `\Spies\GlobalSpies::call_original_global_function` due to lack of redefinition](https://github.com/sirbrillig/spies/blob/master/src/Spies/GlobalSpies.php#L105) and thus loop is not created.

Example code:

```php
<?php

class Test_Foo extends PHPUnit\Framework\TestCase {
	function tearDown() {
		\Spies\finish_spying();
	}

	public function test_foo_1() {
		$bar_spy = \Spies\get_spy_for( 'bar' ); // stubbing previously undefined global function
		$ret = bar();
		$this->assertTrue(
			$bar_spy->was_called_times( 1 )
		);
		$this->assertEquals( $this::$ret, $ret );
	}

	public function test_foo_2() {
		$bar_spy = \Spies\get_spy_for( 'bar' );  // re-stubbing
		$ret = bar();
		$this->assertTrue(
			$bar_spy->was_called_times( 1 )
		);
		$this->assertEquals( $this::$ret, $ret );
	}
}
```
Invocation of the second `bar` above will create an infinite loop.

Note: this case doesn't generate if the global function exists beforehand or if the re-stub has a return value. In those cases, `\Spies\GlobalSpies::call_original_global_function` is not called and there is no loop.

### PR DESCRIPTION
**Pull request by mehamasum**
Fixes https://github.com/sirbrillig/spies/issues/31

(I know this repo has not been developed in long time but your feedback is appreciated. :smile: )

The first commit adds tests for https://github.com/sirbrillig/spies/pull/28, to make sure there is no regression.

The second and third commit respectively tests and fixes the problem described in https://github.com/sirbrillig/spies/issues/31.

Without the change in the third commit, the tests would get stuck in the infinite loop or cause `Allowed Memory Exhausted` error.

### Testing instructions
Please run tests with a memory limit (i.e. `vendor/bin/phpunit -d memory_limit=24M`) so that tests are not stuck in a loop for a long time.

<details>
  <summary>Apply this patch (to remove @runTestsInSeparateProcesses and add finish_spying)</summary>
  
  ```
  Index: tests/StubTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/StubTest.php b/tests/StubTest.php
--- a/tests/StubTest.php	
+++ b/tests/StubTest.php	
@@ -1,9 +1,11 @@
 <?php
 
-/**
- * @runTestsInSeparateProcesses
- */
+
 class StubTest extends PHPUnit_Framework_TestCase {
+	public function tearDown() {
+		\Spies\finish_spying();
+	}
+
 	public function test_mock_function_returns_a_spy() {
 		$stub = \Spies\mock_function( 'test_stub' );
 		$this->assertTrue( $stub instanceof \Spies\Spy );
Index: tests/SpyTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/SpyTest.php b/tests/SpyTest.php
--- a/tests/SpyTest.php	
+++ b/tests/SpyTest.php	
@@ -7,10 +7,12 @@
 	return $val;
 }
 
-/**
- * @runTestsInSeparateProcesses
- */
+
 class SpyTest extends PHPUnit_Framework_TestCase {
+	public function tearDown() {
+		\Spies\finish_spying();
+	}
+
 	public function test_spy_was_called_returns_true_if_called() {
 		$spy = new \Spies\Spy();
 		$spy();
Index: tests/MockObjectTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/MockObjectTest.php b/tests/MockObjectTest.php
--- a/tests/MockObjectTest.php	
+++ b/tests/MockObjectTest.php	
@@ -14,10 +14,12 @@
 	}
 }
 
-/**
- * @runTestsInSeparateProcesses
- */
+
 class MockObjectTest extends PHPUnit_Framework_TestCase {
+	public function tearDown() {
+		\Spies\finish_spying();
+	}
+
 	public function test_mock_object_returns_mock_object() {
 		$mock = \Spies\mock_object();
 		$this->assertTrue( $mock instanceof \Spies\MockObject );

  ```
</details>

- on `master`, comment out `self::$global_functions = [];` in `src/Spies/GlobalSpies.php` (one liner change from https://github.com/sirbrillig/spies/pull/28)  run the tests; some tests fail, so there could be come problems with `finish_spying` (indicates the problem that PR tries to fix)
- on `master`, revert the one liner change and run tests again: tests get memory limit error, so the infinite loop is possibly introduced (indicates the problem this PR is trying to fix)
- checkout first commit of this branch, run the tests: no changes, so both issues remain but there is test for first issue now
- checkout second commit of this branch, run the tests: no changes, so both issues remain but there is test for both issues now
- checkout third commit of this branch, run the tests: tests pass, so both issues should be resolved by now

### CODE REVIEW

File: src/Spies/GlobalSpies.php [hunk 1]
```diff
@@ -121,10 +131,7 @@ private static function replace_global_function( $function_name ) {
 		}
 		self::$redefined_functions[ $function_name ] = \Patchwork\redefine( $function_name, function() use ( $function_name ) {
 			$value = \Spies\GlobalSpies::handle_call_for( $function_name, func_get_args(), [ 'return_falsey_objects' => true ] );
-			if ( isset( $value ) ) {
-				return self::filter_return_for( $value );
-			}
-			return \Patchwork\relay();
```
Comment (mehamasum) [inline src/Spies/GlobalSpies.php:127]: I am not entirely sure why we need to relay here.
I did not intend to change this part (this is not directly part of the fix) but without this, the call count of a stubbed method gets increased. For example:

```
1) ConsecutiveExecutionTest::test__undefined_function___can_be_restubbed
Failed asserting that 2 matches expected 1.
```

**Comment by mehamasum** [comment]
Some tests are not running for me locally, I might have broken a few things. I will continue working on that.

File: src/Spies/GlobalSpies.php [hunk 2]
```diff
@@ -105,6 +109,12 @@ public static function call_original_global_function( $function_name, $args ) {
 		if ( ! isset( self::$redefined_functions[ $function_name ] ) ) {
 			return;
 		}
+
+		// skip if the function did not exist before (i.e. had no "original" function)
+		if ( isset( self::$generated_functions[ $function_name ] )) {
```
Comment (sirbrillig) [inline src/Spies/GlobalSpies.php]: Very minor spacing here (if only we had prettier):
```suggestion
		if ( isset( self::$generated_functions[ $function_name ] ) ) {
```

File: src/Spies/GlobalSpies.php (hunk 1 above)
Comment (sirbrillig) [reply src/Spies/GlobalSpies.php:127]: I'm having trouble finding the docs for this version of Patchwork to remember what this did exactly, but I assume the reason we relay here is so that a global function could be spied upon without needing to also mock it. Something like:

```php
function getName() {
  return 'human';
}
//...
$spy = \Spies\get_spy_for( 'getName' );
$name = getName();
$this->assertSame( 'human', $name );
$this->assertTrue( $spy->was_called() );
```

If this is removed, won't that prevent the original global function from being called?

Although, I see that you have a test `test__global_function__can_be_restubbed` which covers this case, and it passes. How does that work?

Comment (sirbrillig) [reply src/Spies/GlobalSpies.php:127]: Ok, I followed the code and it looks like when `GlobalSpies::handle_call_for()` is called, it eventually calls `GlobalSpies::call_original_global_function()` thanks to `Spy->get_return_for()`, which uses `\Patchwork\restore()` to call the function before recreating the mock with `GlobalSpies::replace_global_function()`.

That explains why this works. Maybe that wasn't always the case which would explain the relay? Or maybe there's another edge case we're not considering here. 🤷

**Review by sirbrillig — APPROVED** [review]
Looks fine to me! Thank you so much for adding tests.
